### PR TITLE
[Dependency Injection] Add autowiring types for aliases

### PR DESCRIPTION
**Pull request by hason**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #18040 |
| License | MIT |
| Doc PR |  |

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Alias.php [hunk 1]
```diff
@@ -47,6 +49,74 @@ public function setPublic($boolean)
...
+    {
+        return isset($this->autowiringTypes[$type]);
+    }
+
+    /**
+     * Adds a type that will default to this definition.
+     *
+     * @param string $type
+     *
+     * @return Definition The current instance
```
Comment (chalasr) [inline src/Symfony/Component/DependencyInjection/Alias.php]:
```diff
- @return Definition
+ @return Alias
```

Same below

**Comment by ogizanagi** [comment]
Is it a good idea to use aliases for setting the autowiring type ?
I don't really have arguments against, but it feels strange to me to rely on aliases to "add something to the original definition". :/

**Comment by hason** [comment]
@ogizanagi see https://github.com/symfony/symfony/pull/18040#issuecomment-195779858

**Comment by stof** [comment]
@hason the idea is to say ``Use the "router" service when someone wants a UrlGeneratorInterface`` instead of saying ``Use the private "router.default" service``. This way, the autowired services would benefit from alias replacements

File: src/Symfony/Component/DependencyInjection/Alias.php (hunk 1 above)
Comment (hason) [inline src/Symfony/Component/DependencyInjection/Alias.php]: Fixed

**Comment by hason** [comment]
Oh no! This doesn't work. The autowiring types are lost after each change of alias definition.

**Comment by nicolas-grekas** [comment]
Did someone try using a ChdilDefinition?
```yaml
services:
   foo_type:
       public: false
       parent: foo
       autowiring_types: [ FooInterface ]
```

**Pull request closed without merging**